libcudacxx/include/cuda/std/__bit/popcount.h

```diff
@@ -22,65 +22,43 @@
 #endif // no system header
 
 #include <cuda/std/__bit/popc.h>
-#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__concepts/concept_macros.h>
+#include <cuda/std/__type_traits/conditional.h>
 #include <cuda/std/__type_traits/is_unsigned_integer.h>
 #include <cuda/std/cstdint>
 #include <cuda/std/limits>
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<sizeof(_Tp) <= sizeof(uint32_t), int>
-__popcount_dispatch(_Tp __t) noexcept
-{
-  return __cccl_popc(static_cast<uint32_t>(__t));
-}
-
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<sizeof(_Tp) == sizeof(uint64_t), int>
-__popcount_dispatch(_Tp __t) noexcept
+_CCCL_TEMPLATE(class _Tp)
+_CCCL_REQUIRES(_CCCL_TRAIT(_CUDA_VSTD::__cccl_is_unsigned_integer, _Tp) _CCCL_AND(sizeof(_Tp) <= sizeof(uint64_t)))
+_LIBCUDACXX_HIDE_FROM_ABI constexpr int __popcount(_Tp __t) noexcept
 {
-  return __cccl_popc(static_cast<uint64_t>(__t));
+  using _Sp = _If<sizeof(_Tp) <= sizeof(uint32_t), uint32_t, uint64_t>;
+  return _CUDA_VSTD::__cccl_popc(static_cast<_Sp>(__t));
 }
 
-template <typename _Tp, int _St = sizeof(_Tp) / sizeof(uint64_t)>
-struct __popcount_rsh_impl
-{
-  static _LIBCUDACXX_HIDE_FROM_ABI constexpr int __count(_Tp __t)
-  {
-    return __popcount_rsh_impl<_Tp, _St - 1>::__count(__t >> numeric_limits<uint64_t>::digits)
-         + __cccl_popc(static_cast<uint64_t>(__t));
-  }
-};
-
-template <typename _Tp>
-struct __popcount_rsh_impl<_Tp, 1>
+_CCCL_TEMPLATE(class _Tp)
+_CCCL_REQUIRES(_CCCL_TRAIT(_CUDA_VSTD::__cccl_is_unsigned_integer, _Tp) _CCCL_AND(sizeof(_Tp) > sizeof(uint64_t)))
+_LIBCUDACXX_HIDE_FROM_ABI constexpr int __popcount(_Tp __t) noexcept
 {
-  static _LIBCUDACXX_HIDE_FROM_ABI constexpr int __count(_Tp __t)
+  constexpr int _Ratio = sizeof(_Tp) / sizeof(uint64_t);
+  int __count          = 0;
+  for (int __i = 0; __i < _Ratio; ++__i)
```

Comment:
I believe _Ratio and __i should be size_t, we might get warnings otherwise
```suggestion
  constexpr size_t _Ratio = sizeof(_Tp) / sizeof(uint64_t);
  int __count          = 0;
  for (size_t __i = 0; __i != _Ratio; ++__i)
```

Reply:
this is the only place where changing from `int` to `size_t` could make sense. In all other case, it just adds additional overhead in CUDA

Reply:
We can happily stay with 32 bit integrals, but then we need to make sure that we do not break due to conversion warnings

Reply:
no warning with current compilers